ci(build): add windows arm64 build

rbqvq opened the pull request:
## Description
Since [GitHub add Windows 11 Arm64 runners](https://github.blog/changelog/2025-04-14-windows-arm64-hosted-runners-now-available-in-public-preview/), we can prepare Sunshine for ARM64

Let's add windows arm64 builds

### Screenshot
None

### Issues Fixed or Closed
None

## Type of Change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Dependency update (updates to dependencies)
- [ ] Documentation update (changes to documentation)
- [ ] Repository update (changes to repository files, e.g. `.github/...`)

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated the in code docstring/documentation-blocks for new or existing methods/components

---

This Pull Request is Work In Progress, I need more help.

## Changes
1. Add `windows-11-arm64` for GitHub Actions
2. Windows 11 Arm64 runners don't like x86_64 runners have two disk, they're only one disk `C:`, replace hardcoded path `/d/a/_temp/` to `${{ runner.temp }}`
3. Update CMakeFiles.txt to support `CLANGARM64` MSYSTEM

## Known Issues
1. Fix https://github.com/LizardByte/build-deps/pull/415 first
2. FFmpeg build failed on CI, x264 can not found by pkg-config
3. I can not build on my Snapdragon Compute Platform, it said `ERROR: cuda requested, but not all dependencies are satisfied: ffnvcodec`, but CI succeed

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/LizardByte/build-deps?pullRequest=420) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/LizardByte/build-deps?pullRequest=420) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/LizardByte/build-deps?pullRequest=420) it.</sub>

ReenigneArcher commented:
Thanks for the PR. Can you re-enable the other matrix options?

Also, I assume this will fail until https://github.com/LizardByte/build-deps/issues/414 is resolved.

rbqvq commented:
> Thanks for the PR. Can you re-enable the other matrix options?
> 
> Also, I assume this will fail until #414 is resolved.

Its just make CI test be faster, GitHub Actions cancel workflow is too slow, that why I disable it.
After Windows Arm64 CI passed, I will re-enable it.

---

Yes, It will failed.
I will subscribe #414 and #415, thank for tips

ReenigneArcher commented on .github/workflows/ci.yml:
```diff
@@ -268,27 +277,28 @@ jobs:
         # if a dependency needs to be pinned, see https://github.com/LizardByte/build-deps/pull/186
         if: runner.os == 'Windows'
-        uses: msys2/setup-msys2@v2
+        uses: msys2/setup-msys2@v2.26.0
```
please leave on v2... the action updates this tag on each release

looks like this, in combination with the disabled update is how you're pinning cmake?

We can pin it by doing something like this in the comment

```
# if a dependency needs to be pinned, see https://github.com/LizardByte/build-deps/pull/186
```

Or, I recently did this in Sunshine: https://github.com/LizardByte/Sunshine/pull/3831

rbqvq replied:
Disable update and use older MSYS version, it bundle CMake with v3.

I just test it, its works fine in local and CI

ReenigneArcher replied:
Sometimes there are nuances about compiling Sunshine with the same versions of dependencies used here, otherwise the build and/or link will fail in Sunshine.

rbqvq replied:
I think MSYS2 should have a built-in package database. If we don't update (don't refresh the database), then it should point to the old version.

rbqvq replied:
To be honest, I don't really agree with what you're doing.

The method of your pin, if the package of the pin has dependencies (mutual dependencies or many conflicts)
You'll have to suffer.

I think pinng the msys2 version and disabling updates ensures dependency integrity in all environments. We test msys2 every time it is updated, and then update the ci for sunshine and build-deps at the same time.

rbqvq replied:
If you turn on update and use v2 tag, the dependent package version used by build-deps may not be consistent with sunshine.

rbqvq replied:
The same is true of Linux, of course, but Ubuntu is not as radical as Arch Linux updates. (MSYS2 use ArchLinux User Repository | AUR)

rbqvq commented:
A Good news for everyone who care this.

All Works is Done!!!

- Local build: `Passed`
- CI build: `Passed`

Today progress
- Fixed x264 build fails
- Fixed FFmpeg build fails
- Add document for Windows Arm64 build guide

See commits for details

In addition, I pinned the CMake version of CI to 3.31.2, which can temporarily solve the problem of Windows CI build failure.

Thanks to everyone who reads here, have a nice day!

ReenigneArcher commented on .github/workflows/ci.yml:
```diff
@@ -268,27 +277,28 @@ jobs:
         if: runner.os == 'Windows'
         run: |
-          cat /d/a/_temp/setup-msys2/msys2.CMD
+          cat "${{ runner.temp }}/setup-msys2/msys2.CMD"
```
```suggestion
          cat "${{ runner.temp }}/setup-msys2/msys2.CMD"
          echo "MSYS2_PATH=${{ runner.temp }}/msys64" >> $GITHUB_ENV
```

This can simplify the detection logic in the CMakeLists.txt. Then just check for this env variable and if it's not set go with the default. It would also self built versions to specify the path without knowing about the Github runner env variables.

rbqvq replied:
Done

ReenigneArcher replied:
```cmake
set(MSYS2_PATH "C:/msys64")
if(WIN32 AND DEFINED ENV{RUNNER_TEMP})
    set(MSYS2_PATH "$ENV{RUNNER_TEMP}/msys64")
    message(STATUS "Detected CI, get msys2 path from environment: $ENV{RUNNER_TEMP}/msys64")
endif()
```

This will also need to be updated to read from the MSYS2_PATH env variable now.

rbqvq replied:
Done :D

rbqvq replied:
MSYS2_PATH was taken by MSYS2 shell.
I replace it to MSYS2_ROOT

ReenigneArcher commented on CMakeLists.txt:
```diff
@@ -71,11 +77,17 @@ if(WIN32)
     set(MSYSTEM "UCRT64")
-    find_file(MSYS2_EXECUTABLE NAMES msys2_shell.cmd REQUIRED HINTS D:/a/_temp/msys64 C:/msys64)
+    set(MSYS2_ARG "-ucrt64")
+    if(CMAKE_HOST_SYSTEM_PROCESSOR MATCHES "ARM64")
```
```suggestion
    if(CMAKE_SYSTEM_PROCESSOR MATCHES "ARM64")
```

For cross compiling (not sure if that will be possible) I think this is probably better

Edit: Oh, looks like you answered that in the readme, but probably still go with this change as it will match the host architecture when not cross compiling.

rbqvq replied:
I think no way to compile arm64 version on x64 windows, but i'm not sure arm64 windows build x64 version

rbqvq replied:
Anyway, I don't recommend compiling x64 packages under arm64, the ucrt64 component is x64 native binary, and the binary translation effect is very bad.

rbqvq replied:
Done. Also add CLANGARM64 check, on who want build win-arm64 target on non-arm64 platform, it will throw a fatal error now.

rbqvq commented:
Here is CI artifacts
We can try build Sunshine for Windows Arm64 now

https://github.com/rbqvq/build-deps/actions/runs/14957436940/artifacts/3101752802

ReenigneArcher commented on cmake/ffmpeg/ffmpeg.cmake:
```diff
@@ -52,17 +52,35 @@ list(APPEND FFMPEG_EXTRA_CONFIGURE
+    #
+    # We must disable CUDA and NVENC until following issues is resolved
+    #
+    # https://github.com/FFmpeg/FFmpeg/blob/4e5523c98597a417eb43555933b1075d18ec5f8b/configure#L7443
```
Do we need do exclude the flags or will ffmpeg automatically disable them when it sees the architecture is arm?

rbqvq replied:
FFmpeg will disable ffnvcodec (But I edit FFmpeg configure, it build find)
If not disable it, configure will throw `ERROR: cuda requested, but not all dependencies are satisfied: ffnvcodec`

Maybe I need send FFmpeg a PR.

ReenigneArcher replied:
> Maybe I need send FFmpeg a PR.

Not necessary, I just wanted to double check.

ReenigneArcher commented on README.md:
```diff
@@ -121,6 +121,12 @@ brew install \
+
+> For example, if you want to compile the arm64 version, you must use Windows On Arm to compile.
+
+- x86_64 / amd64
```
```suggestion
##### x86_64 / amd64
```

ReenigneArcher commented on README.md:
````diff
@@ -139,6 +145,30 @@ pacman -S \
     mingw-w64-ucrt-x86_64-onevpl
 ```
 
+- aarch64 / arm64
````
```suggestion
##### aarch64 / arm64
```

ReenigneArcher commented on README.md:
```diff
@@ -121,6 +121,12 @@ brew install \
 
 #### Windows
 
+⚠ You must use the same CPU architecture as the compilation target under Windows!
```
Is this actually true? Reading through some comments on one of the issues you linked it seems that msys2 is running under emulation when installed to an arm64 machine, so perhaps we don't even need to use a native arm system?

rbqvq replied:
MSYS2 contains x86_64 components, only `clangarm64` are native arm64 binary (that can not opened on x86 windows)

ReenigneArcher replied:
Got it, thanks!

ReenigneArcher reviewed (commented):
Thanks for updating the PR!

It looks pretty good but I do have a few comments and questions.

ReenigneArcher commented:
P.S. can you please sign the CLA?

https://github.com/LizardByte/build-deps/pull/420#issuecomment-2833534326

rbqvq commented:
> P.S. can you please sign the CLA?
> 
> [#420 (comment)](https://github.com/LizardByte/build-deps/pull/420#issuecomment-2833534326)

Done (~~I remember I signed it when I sent PR~~)

rbqvq commented:
Thanks
